### ISSUE TITLE
Allow RE to execute sub plans in parallel

### ISSUE THREAD

**Issue by whs92** [bluesky/bluesky#1652]
Currently the run engine does one task at a time. There are use cases where we would like to ask the run engine to "run this plan/these plans and tell me when you are done". Example use cases are:

- Running alignment or calibration procedures which are independent of each other while also preparing a robot to select a new sample
- Change beamline energy while a sample robot get's a new sample
- Move/Spin a sample in a given trajectory while running a measurement technique
- Arm multiple detectors at the same time
- https://bliss.gitlab-pages.esrf.fr/bliss/1.10.x/motion_hook.html 

In each of these cases there are different ways to solve the problem. Execution could be handled in Ophyd, or in EPICS/Tango/etc layer or in the device itself. This is discussed in https://github.com/bluesky/ophyd/issues/1121#issuecomment-1579146876. @tacaswell touched on the problem some time ago in https://github.com/bluesky/bluesky/pull/541

Suggested reading by @tacaswell: https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/

## Proposed Changes

- A new message added to the RE which takes as arguments a plan and a set of verbs that are allowed to be used. This allows us to have some control over what plans are allowed and not. `async def _run_sub(self, plan, verbs)` should return a status object
- RE should be modified to be able to run sub_plans in parrallel in a way that is backwards compatible.

**Comment by DominicOram** [comment]
As discussed, for a lot of these cases we can get away with a very small subset of verbs that the `_run_subs` is allowed. I suggest if whoever does the implementation finds that it's more complex to allow the verbs that interact with the Event system they spin those into a new issue

**Comment by tacaswell** [comment]
The Msg should return a status object to the calling plan and it should particpate in the same wait logic that set and trigger do so you can write things like:

```python
def my_plan():
   ...
   yield from run_sub_plan(adjust_energy(), group='setup')
   yield from run_sub_plan(change_sample_with_robot(), group='setup')
   yield from bps.set(detector_pos, 5, group='setup')
   yield from bps.wait(group='setup')
   ...
   yield from bp.count(...)
```

other design questions:
  - should it take a called plan (no need to get involved in args/kwargs, but might lead to sticky situations with rewind), insist on taking a no parameter callable (still no arg/kwargs forwarding, and gives the option of not doing "rewind" in the sub-plans by saying it will just run the sub-plan again if replayed), or do the full args/kwargs forwarding.  asyncio went with the "parameter less" function option (and says "use functools.partial if you want to do the args/kwargs binding") so that is probably the right choice here.
  - cleaning up any running tasks on return is straight forward (we already have logic to call `stop` on everything we touched so it should just be extending that), but we may want to add a way to cancel these explicitly or rely on `try...except` to get the scoping right (so only kill them if something else goes wrong / we are interputed).  I think we do want a clean way to cancel them so that you can write infinite subplans and then gracefully clean them up.
  - how to handle failures in the sub-plans, both from a "how do we get a good traceback to the user" and "how do we abort everything else point of view".  For the second I think we can follow the model for how we handle failures on `set`
  - lament that we do not have `gwith` to complete the set of `with` and `awith` (not really a design choice.  The problem is you can not write a context manager that will yield on cleanup.  I had some conversations a long time ago with core developers and they did not think we could sell Guido on adding `gwith` from a completionist point of view and did not think our use case was broad enough to warrant a new keyword in the language. Maybe the landscape on this is different now).  The reason it is worth lamenting this here is that may prevent us from using context managers as nicely as you can in the trio context.

**Comment by untzag** [comment]
For a while I've been playing with the idea of a special plan that accepts other plans and runs all of them simultaneously by "intermixing" the messages fed to the RE. Finally sat down to try writing such a plan. I think I like this pattern, sharing here in the unlikely case that others find it useful.

```python
import uuid
import bluesky
import collections
from dataclasses import dataclass, field
from bluesky import Msg
from bluesky import plan_stubs as bps
import ophyd
from ophyd import sim, status


@dataclass
class StatefulPlan:
    plan: object
    statuses: list = field(default_factory=list)
    waiting: bool = False


class WaitLocalizer(object):

    def __init__(self, *args):
        self.sps = list()
        for plan in args:
            sp = StatefulPlan(plan=plan)
            self.sps.append(sp)
        self.uuid = uuid.uuid1()

    def __iter__(self):
        while True:
            for sp_index, sp in enumerate(self.sps):
                if sp.waiting:
                    continue
                while True:
                    try:
                        msg = next(sp.plan)
                    except StopIteration:
                        self.sps.pop(sp_index)
                        break
                    if msg.command == "wait":
                        sp.waiting = True
                        break
                    print(msg.command, msg.obj.name, msg.args)
                    ret = yield(msg)
                    if isinstance(ret, bluesky.protocols.Status):
                        sp.statuses.append(ret)
                        ret.add_callback(self.status_callback)
            if len(self.sps) == 0:
                print("returning")
                return
            yield Msg("trigger", self, group=self.uuid)
            print("waiting")
            yield Msg("wait", group=self.uuid)

    def status_callback(self, *args, **kwargs):
        for sp in self.sps:
            for index, status in enumerate(sp.statuses):
                if status.done:
                    sp.statuses.pop(index)
            if len(sp.statuses) == 0:
                sp.waiting = False
        try:
            self.status.set_finished()
        except ophyd.utils.errors.InvalidState:
            pass

    def trigger(self):
        self.status = status.Status()
        return self.status


# slow down motors
sim.motor1.delay = 2
sim.motor2.delay = 0.321

def plan1():
    for x in range(-10, 10, 2):
        yield from bps.mv(sim.motor1, x)

def plan2():
    for x in range(0, 100, 5):
        yield from bps.mv(sim.motor2, x)

def plan():
    yield Msg("open_run")
    yield from WaitLocalizer(plan1(), plan2())
    yield Msg("close_run")


RE = bluesky.RunEngine()
RE(plan())
```

The basic idea is
- pass through messages from the subplans to the RE and collect status objects as you see them---keep track of witch statuses came from which subplans
- if you see a wait message, intercept it---don't pass it to the RE
- keep passing messages until all the plans are in a waiting state, then have the RE wait on a special status that you control
- once any of the subplan status objects you've collected finish, release special status and RE will continue consuming messages from any plans that aren't waiting

It does work, for a limited set of plans... Mostly a personal learning exercise... To work properly, `WaitLocalizer` would need to grow proper support for `sleep` and respect for `groups` of status objects. I'm also pretty sure there's a race condition if the subplan statuses finish too quickly, which needs to be accounted for.

**Comment by tacaswell** [comment]
@whs92 Yes, that is the gist of it.  Once we start the `N` sub-plans we need to make sure that they either finish or are cancelled no matter what else happens and getting that right is subtle.

For context, a simpler version of this problem came up with the caproto servers which have a bunch of background tasks that need to be canceled on exit.  With trio I think we got it right on the first try, with asyncio it took us four or five tries to get it right.

**Comment by danielballan** [comment]
Yeah that article has a lot in it, it took me a couple passes to follow it. Specifically I think the lesson for us is, asyncio makes it easy to do cancellation wrong.

**Comment by clintonroy** [comment]
One potential use case that might help; we're running a plan at the moment that triggers does a count([detector, beamstop], 100) basically, at the same time we want to monitor a spectrometer, so we use monitor_during_wrapper(count[detector, beamstop], 100), [spectro]).

However, we need to trigger and read from the spectro in a different cadence to the detector, so atm I asyncio.create_task a background task to trigger the spectro every 0.x seconds, while the main task is monitoring the spectro.

Ideally, I could just have a separate plan that triggers and reads the spectro, that runs side by the side the count([detector, beamstop]) plan.

**Comment by dperl-dls** [comment]
I'm going to have a go at implementing some very basic version of this.

I think a basic list of allowed verbs could be:
```python
ALLOWED_VERBS = {
    "null", "stop", "set", "trigger", "sleep", "wait", "prepare",
    "collect", "kickoff", "complete", "stage", "unstage",
}
```

Can anyone identify any obvious serious issues with running these in the background? The idea is that the "main thread" would still be able to interact with the event system, and this still ought to cover all the use cases covered in the issue description.

I'm also interested in what people think should happen if a `close_run` happens (for an outermost run) while any subplans are running? Should we:
- a. error, tell the user they must explicitly wait?
- b. block and wait?
- c. cancel them and warn?
- d. something else?

**Comment by tacaswell** [comment]
My thought had been to do with with tasks (so each sub-plan gets its own task and can context switch at any await).  This might allow too much switching, but only context switching on `wait` (and `sleep`?) seems like too little.  

The simplest thing on rewinding is to ban it while sub-plans are running (if you cache the "launch a subplan" message _and_ the messages the subplan generates on rewind you will duplicate them.   We have a preview of some of this complexity comes from "interlace all the plans in the main `_run`" from suspenders (which have the ability to add a plan to the stack of plans) and I'm worried about it!

Can a sub-plan launch sub-plans?  If so, if you cancel a sub-plan how do you make sure it and its children (but not any others) get canceled?

Given the amount of state `_run` holds I'm worried about the complexity if making each of those an array when you get the same effect (via a new `locals()`) from the task-based approach and let Python do the book-keeping for us.

If something like what @untzag proposed is sufficient, then I think that should go into `bluesky.preproessors` more-or-less as is without having to touch `_run`.

**Comment by tacaswell** [comment]
As for what to do about `close_run`, consider something like:

```python

def jogger(motor, start, stop):
    for target in itertools.cycle([start, stop]):
        yield from bps.mv(motor, target)
        yield from bps.checkpoint()

def main_plan():
    token = yield from run_subplan(jogger(motor, start, stop))
    for j in range(15):
         yield from bp.count(...)
    yield from abort_subplat(token)
```

should be possible as should

```python
def background(det, delay=5):
    while True:
        yield from bps.trigger_and_read([det], stream='background')
        yield from bps.checkpoint()
        yield from bps.sleep(delay)


def main_plan():
    yield from bps.open_run()
    token = yield from run_plan(background(det1))
    for j in range(15):
        yield from bps.mv(mtr, j)
        yield from trigger_and_read([other, dets])
        yield from bps.checkpoint()
    yield from abort_run(token)
    yield from close_run()
```

I suspect that we need to be careful about managing the bundlers carefully so that if the run is closed, the bundlers go away and the second case would fail "naturally".

I think on `_run` exit we have to make everything stop, but that is different than `close_run`.

**Comment by dperl-dls** [comment]
> My thought had been to do with with tasks (so each sub-plan gets its own task and can context switch at any await). 

I don't really understand what would be executing on the messages in this case, do you mean creating a new `RE` instance and creating a task out of `partial(RE, subplan_generator, is_subplan=True)`?

> The simplest thing on rewinding is to ban it while sub-plans are running 

Yes, that was what I meant

> Can a sub-plan launch sub-plans? If so, if you cancel a sub-plan how do you make sure it and its children (but not any others) get canceled?

No, I wasn't going to allow `run_subplans` in subplans - no use case we have imagined so far would need it, and as you say it introduces a lot of extra complexity

> Given the amount of state _run holds I'm worried about the complexity if making each of those an array when you get the same effect (via a new locals()) from the task-based approach and let Python do the book-keeping for us.

sure, I agree; happy to try out both approaches and compare them - but I wasn't going to make all of the `RE` state into arrays, since the subplans don't need a vast majority of it, just add a new class to manage the minimal state that they do need. the only change I'd hoped to make to `_run` was `where to pull the next message from`

**Comment by dperl-dls** [comment]
After discussing this with @DominicOram @callumforrester etc. we're convinced that https://github.com/bluesky/ophyd/issues/1121#issuecomment-1579146876 / https://github.com/bluesky/bluesky/issues/1654 is indeed the better way to approach this, so I will have a look at doing that first

**Comment by tacaswell** [comment]
> no use case we have imagined so far would need it

Say you have a subplan to do some non-trivial physical reconfiguration of the beamline (move a sample through a load lock, move a sample from a sample library to being mounted for measurement, moving a bunch of mirrors + optics to change which end station is going to get beam, moving optics + detectors to change what technique is being used etc) and you want to be able to factor the steps of that nicely.

```python
def subplan_to_change_optics():
    ...

def subplan_to_change_detectors():
    ...

def subplan_change_technique():
    yield from run_subplan(subplan_to_change_optics)
    yield from run_subplan(subplan_to_change_detectors)
    yield from bps.wait()

def load_sample(sample):
    ...

def do_measurement(sample):
    yield from run_subplan(load_sample, sample)  # however we spell passing parameters through
    yield from run_subplan(subplan_change_technique)
    yield from bps.wait()
    yield from take_data()
```

There may be things we want to prevent subplans from doing (initially I thought "taking data" would be one, but the use-case from the Australian's needs that) but I suspect that those will be things that are likely to mess with other running (sub) plans (like unregistering a message or callback).   Anything we don't support on the first pass we should at least think through if it could _conceptually_ be supported and make sure we do not make it impossible in the future (even if we do not do that work now).

**Comment by tacaswell** [comment]
I can think of contexts where it both makes sense to have the sub-plans share a bundler cache and where it does not.
